library/ssl_tls13_server.c

```diff
@@ -207,16 +207,42 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
      *
      * Servers MUST NOT use any value greater than 604800 seconds (7 days).
      *
+     * `ssl->conf->f_ticket_parse()` might return an illegal `ticket_lifetime`.
+     */
+    if (session->ticket_lifetime >
+        MBEDTLS_SSL_TLS1_3_MAX_ALLOWED_TICKET_LIFETIME) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            3, ("Ticket lifetime exceeds maximum allowed value."));
+        goto exit;
+    }
+
+    /* RFC 8446 section 4.6.1
+     *
+     * The value of zero indicates that the ticket should be discarded
+     * immediately.
+     *
+     * `session->ticket_lifetime == 0` is a internal error for we won't send
+     * this kind of ticket.
+     *
+     */
+    if (session->ticket_lifetime == 0) {
+        MBEDTLS_SSL_DEBUG_MSG(
+            3, ("Ticket lifetime is zero"));
+        goto exit;
+    }
+
+    /*
      * RFC 8446 section 4.2.11.1
      *
      * Clients MUST NOT attempt to use tickets which have ages greater than
      * the "ticket_lifetime" value which was provided with the ticket.
      *
      */
-    if (server_age > MBEDTLS_SSL_TLS1_3_MAX_ALLOWED_TICKET_LIFETIME * 1000) {
+    if (server_age > session->ticket_lifetime * 1000) {
```

Comment:
This is checked by ssl_ticket.c thus no need to check it again here.

Reply:
The documentation is the definition of the type `mbedtls_ssl_ticket_parse_t` in ssl.h. There is no explicit sentence about the check of the lifetime but an MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED error code is defined with its documentation being "Session ticket has expired." It would be good to be more explicit in the documentation for sure and I don't think it would be an API break. @mpg what do you think about that?

Reply:
For TLS 1.3, I think we should add some sentences about max lifetime limitation. With that, I think we can remove many checks in this PR.

Reply:
> What would be the change? I don't think we can add new requirements on users who write their own callbacks. Also, as a general rule, I think it's best if checks required by the protocol are done in library code if possible, not left to callbacks that may be written by users.

That means we should keep the check, right?

Reply:
@mpg , https://github.com/Mbed-TLS/mbedtls/pull/8581#discussion_r1412090648, these checks should keep also, right?

Reply:
Yes, that's my opinion: we should keep checks in `ssl_tls13_server.c`.

Reply:
Technically after the changes in PR #8574 the test in question here is exactly the same as the one in our implementation of the ticket parsing function, namely:
```
if (ticket_age < 0 || ticket_age > ticket_lifetime)
```
Then as I said above, in the documentation of `mbedtls_ssl_ticket_parse_t` in ssl.h an MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED error code is defined with its documentation being "Session ticket has expired." Thus, it does not seem completely out of the picture to me to assume that another implementation would also check the ticket lifetime properly.

Another aspect of this is that with the current definition of `mbedtls_ssl_ticket_parse_t`  and `mbedtls_ssl_ticket_write_t` it does not seem possible to have the ticket lifetime being included in the ticket data (field session->ticket_lifetime above). The TLS stack pass to the write function the data it wants to be included in the ticket as a `const mbedtls_ssl_session *session` without knowing the ticket lifetime and gets in response the encrypted ticket data and the ticket lifetime (but to late ...). We do not have a solution to do the check in question here without breaking the ticket API.

Thus the compromise, I have come up with, where the ticket lifetime is checked properly by the ticket parsing function (#8574) and only the ticket parsing function, not also in the TLS 1.3 server code. I know this is bit border line regarding other implementations of `mbedtls_ssl_ticket_parse_t` but this seems acceptable to me.

Reply:
> Also, as a general rule, I think it's best if checks required by the protocol are done in library code if possible, not left to callbacks that may be written by users.

I agree with @mpg 's point.  We have two possible solution for this issue. For the time being, I think #8547 is better,  although it violates @mpg's point. For a long term, I think #8525 is better.